libs/core/include/core/yaml/document.hpp

```diff
@@ -0,0 +1,177 @@
+
+#include "core/byte_array/byte_array.hpp"
+#include "core/byte_array/const_byte_array.hpp"
+#include "core/byte_array/consumers.hpp"
+#include "core/yaml/exceptions.hpp"
+#include "variant/variant.hpp"
+
+#include <cstddef>
+#include <cstdint>
+#include <vector>
+
+namespace fetch {
+namespace yaml {
+
+class YamlDocument
+{
+public:
+  enum Type
+  {
+    KEYWORD_TRUE     = 0,
+    KEYWORD_FALSE    = 1,
+    KEYWORD_NULL     = 2,
+    KEYWORD_CONTENT  = 3,
+    KEYWORD_INF      = 4,
+    KEYWORD_NEG_INF  = 5,
+    KEYWORD_NAN      = 6,
+    STRING           = 7,
+    STRING_MULTILINE = 8,
+    COMMENT          = 9,
+
+    // TODO(private issue #566): There should be defined more integer and floating point
+    // types here, however they will be relevant only when used with document
+    // SCHEMA defining expected types in the document.
+    NUMBER_INT   = 10,
+    NUMBER_FLOAT = 11,
+    NUMBER_HEX   = 12,
+    NUMBER_OCT   = 13,
+
+    OPEN_OBJECT  = 14,
+    CLOSE_OBJECT = 15,
+    OPEN_ARRAY   = 16,
+    CLOSE_ARRAY  = 17,
+
+    NEW_ENTRY           = 18,
+    NEW_MULTILINE_ENTRY = 19,
+
+    KEY = 20,
+
+    ALIAS           = 21,
+    ALIAS_REFERENCE = 22,
+    TAG             = 23,
+    KEYWORD_TAG     = 24
+  };
+
+  using ByteArray      = byte_array::ByteArray;
+  using ConstByteArray = byte_array::ConstByteArray;
+  using Variant        = variant::Variant;
+
+  explicit YamlDocument(ConstByteArray const &document)
+  {
+    Parse(document);
+  }
+
+  YamlDocument()
+  {
+    // The default of a YamlDocument is to be an empty object
+    variant_ = Variant::Object();
+  }
+
+  YamlDocument(YamlDocument const &) = delete;
+  YamlDocument(YamlDocument &&)      = default;
+  ~YamlDocument()                    = default;
+
+  Variant &operator[](std::size_t i)
+  {
+    return root()[i];
+  }
+
+  Variant const &operator[](std::size_t i) const
+  {
+    return root()[i];
+  }
+
+  Variant &operator[](ConstByteArray const &key)
+  {
+    return root()[key];
+  }
+
+  Variant const &operator[](ConstByteArray const &key) const
+  {
+    return root()[key];
+  }
+
+  Variant &root()
+  {
+    return variant_;
+  }
+
+  Variant const &root() const
+  {
+    return variant_;
+  }
+
+  void Parse(ConstByteArray const &document);
+
+  YamlDocument &operator=(YamlDocument const &) = delete;
+  YamlDocument &operator=(YamlDocument &&) = default;
+
+  bool Has(byte_array::ConstByteArray const &key)
+  {
+    return variant_.Has(key);
+  }
+
+private:
+  constexpr static char const *LOGGING_NAME = "YamlDocument";
+
+  struct YamlObject
+  {
+    Variant *data  = nullptr;
+    uint     ident = 0;
+    uint     line  = 0;
+  };
+
+  struct YamlToken
+  {
+    uint64_t first  = 0;
+    uint64_t second = 0;
+    uint8_t  type   = 0;
+    uint     ident  = 0;
+    uint     line   = 0;
+  };
+
+  void        Tokenise(ConstByteArray const &document);
+  static void ExtractPrimitive(Variant &variant, YamlToken const &token,
```

Comment:
Some of this is defined here, whereas other is defined in the cpp file. Suggest all goes in the cpp file.

Reply:
Agreed.